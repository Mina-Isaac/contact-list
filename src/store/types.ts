export interface NameDetails {
  title: string;
  first: string;
  last: string;
};

interface Location {
  street: { number: number; name: string };
  city: string;
  state: string;
  country: string;
  postcode: string;
}

export interface Contact {
  gender: string;
  name: NameDetails;
  location: Location;
  email: string;
  login: { username: string };
  phone: string;
  cell: string;
  picture: {
    large: string;
    medium: string;
    thumbnail: string;
  };
}

export type APIResponse = {
  results: Contact[];
};
