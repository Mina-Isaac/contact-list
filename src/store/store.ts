import {
  configureStore,
  ThunkAction,
  Action,
  combineReducers,
} from "@reduxjs/toolkit";
import contactsReducer from "./modules/contacts";
import uiReducer from "./modules/ui";

export const rootReducer = combineReducers({
  contacts: contactsReducer,
  ui: uiReducer,
});

export const store = configureStore({
  reducer: rootReducer,
  devTools: process.env.NODE_ENV === "development",
});

export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
