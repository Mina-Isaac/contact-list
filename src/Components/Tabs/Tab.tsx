import React, { useCallback } from "react";
import { useDispatch, useSelector } from "react-redux";
import styled from "styled-components";
import { setSelectedContact, setSelectedTab } from "../../store/modules/ui";
import { RootState } from "../../store/store";

const Wrapper = styled.button`
  /* common styles for all tabs */
  border-radius: 10px 10px 0 0;
  border: var(--borders);
  padding: 3px;
  height: 100%;
  background-color: var(--light-gray);
  display: flex;
  justify-content: space-around;
  align-items: center;
  flex: 1 1 14px;
  color: var(--dark-gray);

  /* disabled styles */
  :disabled {
    color: var(--lighter-gray);
    cursor: not-allowed;
  }

  /* Styles that depend on whether the tab is selected or not */
  border-bottom: ${(props: { selected: boolean }) =>
    props.selected ? "none" : "var(--borders)"};
  background-image: ${({ selected }) =>
    selected
      ? "linear-gradient(to bottom,#969696 -1%,var(--light-gray) 27%)"
      : "none"};
`;

const Letter = styled.span`
  font-weight: bold;
  font-size: 1.5rem;
`;

const Count = styled.span`
  font-size: 0.6rem;
  font-weight: 700;
  align-self: flex-end;
`;

interface Props {
  letter: string;
  count: number;
}

const Tab: React.FC<Props> = ({ letter, count }) => {
  const selectedTab = useSelector((state: RootState) => state.ui.selectedTab);
  const isSelected = letter === selectedTab;
  const dispatch = useDispatch();

  const handleClick = useCallback(() => {
    if (letter !== selectedTab && count > 0) {
      dispatch(setSelectedTab(letter));
      dispatch(setSelectedContact(undefined));
    }
  }, [count, dispatch, letter, selectedTab]);

  return (
    <Wrapper disabled={count === 0} selected={isSelected} onClick={handleClick}>
      <Letter>{letter}</Letter>
      <Count>{count}</Count>
    </Wrapper>
  );
};

export default React.memo(Tab);
