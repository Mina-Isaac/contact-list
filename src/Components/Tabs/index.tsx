import React from "react";
import styled from "styled-components";
import Tab from "./Tab";
import { configJson } from "../../constants";
import { Dictionary } from "lodash/";
import { Contact } from "../../store/types";

const Wrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  margin-bottom: -1px;
`;

interface Props {
  groupedList: Dictionary<Contact[]>;
}

const Tabs: React.FC<Props> = ({ groupedList }) => {
  return (
    <Wrapper>
      {configJson.tabs.map((letter, i) => (
        <Tab key={i} letter={letter} count={groupedList[letter]?.length || 0} />
      ))}
    </Wrapper>
  );
};

export default Tabs;
