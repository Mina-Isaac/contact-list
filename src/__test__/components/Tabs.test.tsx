import React from "react";
import ReactDOM from "react-dom";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import Tabs from "../../Components/Tabs";
import { RootState } from "../../store/store";
import { sampleContactList } from "../fixtures/sampleContactList";
import { groupBy } from "lodash";
const mockStore = configureStore();
let store: any;

describe("Tabs component should render without issues", () => {
  beforeEach(() => {
    const initialState: RootState = {
      contacts: { loading: "fulfilled", error: null, list: sampleContactList },
      ui: { selectedTab: "b", selectedContact: sampleContactList[0] },
    };
    store = mockStore(initialState);
  });
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Provider store={store}>
        <Tabs
          groupedList={groupBy(sampleContactList, (c) =>
            c.name.first[0].toLocaleLowerCase()
          )}
        />
      </Provider>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });
  it("matches snapshot", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Tabs groupedList={groupBy(sampleContactList, (c) =>
            c.name.first[0].toLocaleLowerCase()
          )} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
