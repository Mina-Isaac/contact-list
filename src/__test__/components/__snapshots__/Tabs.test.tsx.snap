// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs component should render without issues matches snapshot 1`] = `
<div
  className="sc-hKgILt fnpaVj"
>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      a
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ fdKsvb"
    disabled={true}
    onClick={[Function]}
    selected={true}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      b
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      c
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={false}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      d
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      1
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      e
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      f
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      g
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      h
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      i
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      j
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      k
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      l
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      m
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      n
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      o
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      p
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      q
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      r
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      s
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      t
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      u
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      v
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      w
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      x
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      y
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
  <button
    className="sc-bdfBwQ hPnPZM"
    disabled={true}
    onClick={[Function]}
    selected={false}
  >
    <span
      className="sc-gsTCUz iwmIdS"
    >
      z
    </span>
    <span
      className="sc-dlfnbm hQdJln"
    >
      0
    </span>
  </button>
</div>
`;
